### DISCUSSION TITLE
how to control the coefficient of binaries generated by the kkt command

### DISCUSSION THREAD

**Discussion by rgzbsn** [yalmip/YALMIP#1282] [Problems using YALMIP]
Hi everyone
I'm trying the kkt command to generate the kkt condition of a MILP model.I've noticed that the kkt command works well except for one row,here is the .lp log of this row:
R32561: 1.052631578947368 C1104 + 1.052631578947368 C1105
   + 1.052631578947368 C1106 + 1.052631578947368 C1107
   + 1.052631578947368 C1108 + 1.052631578947368 C1109
   + 1.052631578947368 C1110 + 1.052631578947368 C1111
   + 1.052631578947368 C1112 + 1.052631578947368 C1113
   + 1.052631578947368 C1114 + 1.052631578947368 C1115
   + 1.052631578947368 C1116 + 1.052631578947368 C1117
   + 1.052631578947368 C1118 + 1.052631578947368 C1119
   + 1.052631578947368 C1120 + 1.052631578947368 C1121
   + 1.052631578947368 C1122 + 1.052631578947368 C1123
   + 1.052631578947368 C1124 + 1.052631578947368 C1125
   + 1.052631578947368 C1126 + 1.052631578947368 C1127 - 0.95 C1128
   - 0.95 C1129 - 0.95 C1130 - 0.95 C1131 - 0.95 C1132 - 0.95 C1133
   - 0.95 C1134 - 0.95 C1135 - 0.95 C1136 - 0.95 C1137 - 0.95 C1138
   - 0.95 C1139 - 0.95 C1140 - 0.95 C1141 - 0.95 C1142 - 0.95 C1143
   - 0.95 C1144 - 0.95 C1145 - 0.95 C1146 - 0.95 C1147 - 0.95 C1148
   - 0.95 C1149 - 0.95 C1150 - 0.95 C1151 + **1.13687e-13 C10598**
   <= 48.12154099193583
It comes that the coefficient of binary(C10598) is very small,which result in a large matrix range of  the model ([1e-13,1e+04]),and thus severly slowing the solving process.However,according to the big-M method,the coefficient of binary could just be 1 or big constant.I'm wondering:
1.how this coefficeint is generated by the kkt command of yalmip?
2.if there is any way to improve this coefficient?
3.if it's hard to convert the coefficient,can I do something to ignore small coefficient in yalmip?
Thanks in advance!

**Comment by johanlofberg** [comment]
A very small coefficient should be no problems for solvers, as they will clean away that if necessary.

A very large coefficient though means kkt wasn't able to derive big-M coefficients, most likely as it didn't manage to derive strong bounds on duals (not surprisingly as it typically is as hard to derive these, as it is to solve the bilevel program to begin with). Hence, you have to add explicit bound on them (details.duals) but then you have to come up with correct bounds. If too small you might cut away the optimal solution, and too large you are still stuck with a numerically poor model

**Reply by rgzbsn** [reply]
Thanks for your advice!I've tried to expand the boundary of duals and it do improve a lot!

**Reply by rgzbsn** [reply]
Hi Johan
Unfortunately the expansion of duals' boundary don't work: as I find a mistake of the model.After I removed the mistake,gurobi struggles to find a solution.The gurobi support team suggests to control the matrix coefficient and RHS range as the tolerance of gurobi is 1e-13,but I'm not sure how to ignore these very small coefficeint.I've tried these measures but it makes no difference:
```
details.A(abs(details.A)<1e-4)=0;
details.Q(abs(details.Q)<1e-4)=0;
details.E(abs(details.E)<1e-4)=0;
details.c(abs(details.c)<1e-4)=0;
details.f(abs(details.f)<1e-4)=0;
```
Thanks in advance!

**Reply by johanlofberg** [reply]
Changing data extracted from the model makes absolutely no difference (unless you use this data to setup the kkt model manually?)

You will have to do something like

```
M = [clean(sdpvar(M(find(is(M,'equality')))),1e-6)==0;
         clean(sdpvar(M(find(is(M,'elementwise')))),1e-6)>=0]
```

I see I have to extend the clean operator...

However, if the solver struggles due to small elements it could easily be that the problem simply is ill-posed. If the small noise is simple and can be removed, then the solver would typically not have an issue anyway. For instance x>=1e-15 would not pose any issue to a solver so removing it makes no difference. On the other hand, if you have 1e-12*x >= y, y>=1 the solver will struggle, but if you remove the small term you are changing the model massively as the model says x has to be huge (and could come from something which actually is modelling an infeasible model where in theory it says that x has to be infinite). 

In other words, clean with caution
